File: backend/app/schemas/market_action.py
"""Pydantic schemas for MarketAction (CRA Art. 35 recall / withdrawal)."""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.models.enums import MarketActionStatus, MarketActionType


class MarketActionCreate(BaseModel):
    product_release_id: UUID
    action_type: MarketActionType
    reason: str = Field(min_length=10)
    affected_scope: str | None = None
    corrective_action: str | None = None
    authority_reference_number: str | None = Field(default=None, max_length=255)
    user_notice_text: str | None = None
    internal_notes: str | None = None


class MarketActionUpdate(BaseModel):
    action_type: MarketActionType | None = None
    status: MarketActionStatus | None = None
    reason: str | None = Field(default=None, min_length=10)
    affected_scope: str | None = None
    corrective_action: str | None = None
    authority_reference_number: str | None = Field(default=None, max_length=255)
    authority_notified_at: datetime | None = None
    user_notice_text: str | None = None
    internal_notes: str | None = None


class ProductReleaseSummary(BaseModel):
    """Minimal release info embedded in MarketActionRead for display purposes."""
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    product_id: UUID
    system_version: int
    user_version: str | None = None

    @computed_field  # type: ignore[misc]
    @property
    def display_version(self) -> str:
        """Human-readable version label: user_version if set, otherwise 'v{system_version}'."""
        return self.user_version or f"v{self.system_version}"


class MarketActionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    product_release_id: UUID
    action_type: MarketActionType
    status: MarketActionStatus
    reason: str
    affected_scope: str | None
    corrective_action: str | None
    authority_reference_number: str | None
    authority_notified_at: datetime | None
    user_notice_text: str | None
    internal_notes: str | None
    product_release: ProductReleaseSummary | None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/market-action.ts
/** TypeScript types for MarketAction (CRA Art. 35 recalls and withdrawals, FR38/FR39). */

export type MarketActionType = "recall" | "withdrawal";

export type MarketActionStatus =
  | "draft"
  | "active"
  | "authority_notified"
  | "closed";

/** Minimal release info embedded in MarketActionRead. */
export interface ProductReleaseSummary {
  id: string;
  product_id: string;
  system_version: number;
  user_version: string | null;
  display_version: string;
}

export interface MarketActionRead {
  id: string;
  product_release_id: string;
  action_type: MarketActionType;
  status: MarketActionStatus;
  reason: string;
  affected_scope: string | null;
  corrective_action: string | null;
  authority_reference_number: string | null;
  authority_notified_at: string | null;
  user_notice_text: string | null;
  internal_notes: string | null;
  product_release: ProductReleaseSummary | null;
  created_at: string;
  updated_at: string;
}

export interface MarketActionCreate {
  product_release_id: string;
  action_type: MarketActionType;
  reason: string;
  affected_scope?: string | null;
  corrective_action?: string | null;
  authority_reference_number?: string | null;
  user_notice_text?: string | null;
  internal_notes?: string | null;
}

export interface MarketActionUpdate {
  action_type?: MarketActionType | null;
  status?: MarketActionStatus | null;
  reason?: string | null;
  affected_scope?: string | null;
  corrective_action?: string | null;
  authority_reference_number?: string | null;
  authority_notified_at?: string | null;
  user_notice_text?: string | null;
  internal_notes?: string | null;
}

export interface AuthorityNotifiedRequest {
  notified_at?: string | null;
}
